### PR TITLE
[LBSE] Add support for 'image' elements

### PR DESCRIPTION
**Pull request by nikolaszimmermann**
#### 4b8776daaa578ce17f6ada73fe7c92704e1945eb
<pre>
[LBSE] Add support for &apos;image&apos; elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=243499">https://bugs.webkit.org/show_bug.cgi?id=243499</a>

Reviewed by Rob Buis.

Rename RenderSVGImage -&gt; LegacyRenderSVGImage, to prepare the LBSE-aware RenderSVGImage introduction.

Re-introduce RenderSVGImage, largely based on RenderImage, unlike LegacyRenderSVGImage.
Supports all features that the legacy engine did, just way more aligned with RenderImage.

Covered by existing tests, fixes 18 additional SVG 1.1 testcases.

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-36-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-39-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-40-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/color-prof-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/coords-viewattr-02-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-blend-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-conv-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-diffuse-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-light-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-specular-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-04-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/render-groups-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/render-groups-03-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-02-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-03-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-04-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-05-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-06-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-07-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-08-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-09-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-10-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-symbol-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-use-01-t-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilitySVGElement.cpp:
(WebCore::AccessibilitySVGElement::accessibilityDescription const):
(WebCore::AccessibilitySVGElement::determineAccessibilityRole):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::renderImageResource):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGImage const):
(WebCore::RenderObject::isSVGImageOrLegacySVGImage const):
* Source/WebCore/rendering/RenderTreeAsText.cpp:
(WebCore::write):
* Source/WebCore/rendering/svg/LegacyRenderSVGImage.cpp: Renamed from Source/WebCore/rendering/svg/RenderSVGImage.cpp.
(WebCore::LegacyRenderSVGImage::LegacyRenderSVGImage):
(WebCore::LegacyRenderSVGImage::willBeDestroyed):
(WebCore::LegacyRenderSVGImage::imageElement const):
(WebCore::LegacyRenderSVGImage::calculateObjectBoundingBox const):
(WebCore::LegacyRenderSVGImage::updateImageViewport):
(WebCore::LegacyRenderSVGImage::layout):
(WebCore::LegacyRenderSVGImage::paint):
(WebCore::LegacyRenderSVGImage::paintForeground):
(WebCore::LegacyRenderSVGImage::invalidateBufferedForeground):
(WebCore::LegacyRenderSVGImage::nodeAtFloatPoint):
(WebCore::LegacyRenderSVGImage::imageChanged):
(WebCore::LegacyRenderSVGImage::addFocusRingRects):
* Source/WebCore/rendering/svg/LegacyRenderSVGImage.h: Renamed from Source/WebCore/rendering/svg/RenderSVGImage.h.
* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.cpp:
(WebCore::isGraphicsElement):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::computeDecoratedBoundingBox const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGImage):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.h:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::SVGRenderingContext::bufferForeground):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createSVGLayerAwareElementSet):
* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::hasSingleSecurityOrigin const):
(WebCore::SVGImageElement::svgAttributeChanged):
(WebCore::SVGImageElement::createElementRenderer):
(WebCore::SVGImageElement::didAttachRenderers):

Canonical link: <a href="https://commits.webkit.org/253793@main">https://commits.webkit.org/253793@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f7ebb051d3720d20ffb602a328f88ac6f0165834

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95958 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149609 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91005 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29480 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25762 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79165 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91058 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92645 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23802 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73851 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23786 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78790 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79061 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66795 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27211 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12899 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27153 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13913 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2679 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36765 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28777 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33190 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/svg/SVGImageElement.cpp
Comment (rwlbuis) [inline Source/WebCore/svg/SVGImageElement.cpp]: Will this not be the most common case in the future and should it not be tested first?

**Review by rwlbuis — COMMENTED** [review]
If the RenderSVGImage -> LegacyRenderSVGImage move is in a separate patch then the RenderSVGImage.cpp will not show up as a complicated diff, but one easy to review blob.

File: Source/WebCore/svg/SVGImageElement.cpp
Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGImageElement.cpp]: Yeah, will adapt, thanks.

**Comment by nikolaszimmermann** [comment]
> If the RenderSVGImage -> LegacyRenderSVGImage move is in a separate patch then the RenderSVGImage.cpp will not show up as a complicated diff, but one easy to review blob.

Indeed, it's a mess to review, hard to see the actual change. Will fix ASAP.

**Comment by nikolaszimmermann** [comment]
@cdumez - Chris, may you have a look at the loader/ related change (it's a rename, but needs your approval as code owner).

File: Source/WebCore/svg/SVGImageElement.cpp
Comment (rwlbuis) [inline Source/WebCore/svg/SVGImageElement.cpp:154]: I guess I will trust you on this, but it is really much simpler if this change is a strict rename. Those can in fact be rubber stamped, more or less.

**Review by rwlbuis — APPROVED** [review]
Looks like you need to regenerate the commit log since no layout tests were changed.

File: Source/WebCore/rendering/svg/RenderSVGImage.cpp
Comment (rwlbuis) [inline Source/WebCore/rendering/svg/RenderSVGImage.cpp]: These two blocks feel like they can be done using one lambda, but maybe it is hard to do here.

File: Source/WebCore/rendering/svg/RenderSVGImage.cpp
Comment (rwlbuis) [inline Source/WebCore/rendering/svg/RenderSVGImage.cpp:252]: Bit of a preference thing, but not sure whether outer parentheses or the whole variable isVisible are needed.

**Review by rwlbuis — APPROVED** [review]

File: Source/WebCore/svg/SVGImageElement.cpp
Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGImageElement.cpp:154]: I will remember this for the next patch, sorry.

File: Source/WebCore/rendering/svg/RenderSVGImage.cpp
Comment (nikolaszimmermann) [reply Source/WebCore/rendering/svg/RenderSVGImage.cpp]: There will be a cleanup patch anyhow for the other RenderSVG* files as well - let's postpone cleanups (the code is copied 1:1 from LegacyRenderSVGImage)

File: Source/WebCore/rendering/svg/RenderSVGImage.cpp
Comment (nikolaszimmermann) [reply Source/WebCore/rendering/svg/RenderSVGImage.cpp:252]: Not strictly, depends on which operator has precedence. I often go for explicit braces....

**Comment by nikolaszimmermann** [comment]
> Looks like you need to regenerate the commit log since no layout tests were changed.
Oh it's in the other patch.... I will squash them now anyhow and rebase since you finished the review.

**Pull request merged**
